Add NVIDIA Parakeet TDT-0.6B v3 ASR

Alex-Wengg opened the pull request:
## Summary
  Introduces NVIDIA's Parakeet TDT-0.6B v3 model for automatic speech recognition that supports multilingual services too, as well as quantizing it to 8 bit integers.

  ## Key Changes
  - New ASR Model: Integrated Parakeet TDT-0.6B v3 CoreML models (encoder, decoder, joint network)
  - TDT Decoder: Changes to the TDT Decoder algo process
    - Predicts both tokens and duration (frame skips) in a single pass
    - Adaptive duration handling for short vs long utterances
  - only greedy decoding algo
  - Expanded token vocabulary sizes of 8193

BrandonWeng commented on Sources/FluidAudio/ASR/AsrManager.swift:
```diff
@@ -310,14 +312,27 @@ public final class AsrManager {
-            jointModel: jointModel!,
-            decoderState: &decoderState
-        )
+        if config.enableBeamSearch {
```
Whats this for?

Alex-Wengg replied:
i was trying to get beamsearch algo to work, we are currently using a simple decoder algo. beam search algo would do something like remember the best 3 past predictions of a encoder / decoder for a bit to pick the best option. but i didn't really found it useful. will remove

BrandonWeng commented on Sources/FluidAudio/ASR/AsrModels.swift:
```diff
@@ -183,6 +218,15 @@ extension AsrModels {
         return config
     }
 
+    /// Create CPU-only configuration for maximum accuracy (matches NeMo notebook)
+    public static func cpuOnlyConfiguration() -> MLModelConfiguration {
+        let config = MLModelConfiguration()
+        config.allowLowPrecisionAccumulationOnGPU = false
+        // CPU-only for Float32 precision (matches the notebook's approach)
+        config.computeUnits = .cpuOnly
+        return config
+    }
+
```
?

BrandonWeng commented on Sources/FluidAudio/ASR/TdtDecoder.swift:
```diff
@@ -114,16 +124,93 @@ internal struct TdtDecoder {
+            let framesProcessed = Float(timeIndices)
+
+            // Calculate dynamic token rate expectation
+            // LibriSpeech averages ~0.03-0.05 tokens per frame
```
Are these numbers just based on vibes?

Alex-Wengg replied:
NVM , i was able to remove it no issue.

Alex-Wengg commented on Sources/FluidAudio/ASR/AsrManager.swift:
```diff
@@ -346,6 +361,32 @@ public final class AsrManager {
         logger.info("Decoder state reset for source: \(String(describing: source))")
     }
 
+    /// Remove leading punctuation that the model sometimes hallucinates
+    /// This is a common issue with v3 multilingual models generating periods/commas at the start
+    private func removeLeadingPunctuation(_ text: String) -> String {
```
i don't think it would impact much but the WER calculation can be over all the place for how it penalize things like a missing punctuation

BrandonWeng commented on Sources/FluidAudio/DownloadUtils.swift:
```diff
@@ -181,20 +185,60 @@ public class DownloadUtils {
-                            NSLocalizedDescriptionKey: "Missing coremldata.bin in model: \(name)",
-                        ])
+                // Check for essential model files based on format
+                if name.hasSuffix(".mlmodelc") {
```
we will only support modelc in the offical package

Alex-Wengg replied:
i will finalize once the quantization and transpose integration are in .

Alex-Wengg commented on Sources/FluidAudio/ASR/AsrTranscription.swift:
```diff
@@ -4,6 +4,41 @@ import OSLog
 extension AsrManager {
 
+    /// Transpose encoder output from [B, H, T] to [B, T, H]
+    private func transposeEncoderOutput(_ input: MLMultiArray) throws -> MLMultiArray {
```
don't you meant the encoder model ? i can prob integrate it to the algo, we did do it for v2 before but the converter script didn't had the transposer merger in mind

Alex-Wengg commented on Sources/FluidAudioCLI/TextNormalizer.swift:
```diff
@@ -99,6 +131,129 @@ struct TextNormalizer {
             normalized = normalized.replacingOccurrences(of: contraction, with: expansion)
         }
 
+        // Normalize British to American spelling variations
+        let spellingVariations = [
+            "counselled": "counseled",
```
yes

BrandonWeng replied:
please leave a comment here for context

BrandonWeng commented:
@claude take a look at the changes. Details around parakeet-tdt-v3 can be found here: https://huggingface.co/nvidia/parakeet-tdt-0.6b-v3

BrandonWeng commented on .github/workflows/asr-benchmark.yml:
```diff
@@ -37,6 +37,37 @@ jobs:
+        run: |
+          MODELS_DIR="$HOME/Library/Application Support/FluidAudio/Models/parakeet-tdt-0.6b-v3-coreml"
+          VOCAB_FILE="$HOME/Library/Application Support/FluidAudio/parakeet_vocab.json"
+
```
I thought the package does this automatically. WE could probably remove now that you've moved it into FLuidInference repo

BrandonWeng commented on Documentation/PARAKEET_V3_JOINT_NETWORK_FIX.md:
```diff
@@ -0,0 +1,156 @@
+# Parakeet v3 Joint Network CoreML Conversion Issue
```
Remove?

Alex-Wengg commented:
> **Claude finished @BrandonWeng's task** —— [View job](https://github.com/FluidInference/FluidAudio/actions/runs/17103610788)
> 
> ### Code Review: NVIDIA Parakeet TDT-0.6B v3 ASR Integration
> **Todo List:**
> 
> * [x]  Read and understand the changed files
> * [x]  Analyze the TDT decoder implementation and optimizations
> * [x]  Review concerns raised in existing review comments
> * [x]  Check integration with existing ASR architecture
> * [x]  Examine text normalization and benchmarking changes
> * [x]  Provide comprehensive code review feedback
> 
> ## Summary
> This PR successfully integrates NVIDIA's Parakeet TDT-0.6B v3 model with strong performance results (7.4%/5.1% WER on LibriSpeech). However, several code quality concerns need to be addressed before merge.
> 
> ## Major Concerns
> ### 1. **Hard-coded Magic Numbers in TdtDecoder (High Priority)**
> The TDT decoder contains concerning hard-coded logic that appears over-optimized for benchmarks:
> 
> ```swift
> // TdtDecoder.swift:162-167 - Very concerning
> if encoderSequenceLength < 30 {
>     actualSkip = 1  // Force single frame skip for short audio
> } else {
>     actualSkip = min(moreDuration, 2)  // Cap at 2 frames for longer audio
> }
> ```
> 
> **Issues:**
> 
> * These thresholds (30, 2) seem arbitrary and benchmark-tuned
> * No documentation explaining the reasoning behind these specific values
> * Could perform poorly on audio lengths near the 30-frame boundary
> * May not generalize to different audio types or speaking patterns
> 
> **Recommendation:** Replace with adaptive logic based on audio characteristics or make configurable parameters.
> 
> ### 2. **Questionable BeamSearch Remnants**
> ```swift
> // AsrManager.swift:24 - Commented out beam search model
> /// Token duration optimization model
> ```
> 
> The code contains remnants of beam search implementation that was removed. This creates confusion about the current decoding strategy.
> 
> **Recommendation:** Clean up all beam search references and clearly document that only greedy decoding is supported.
> 
> ### 3. **Incomplete Integration Issues**
> Several integration concerns raised in reviews:
> 
> * **EncoderWrapper Integration**: `AsrTranscription.swift` appears to duplicate functionality that should be in `EncoderWrapper`
> * **Simple Fallback Pattern**: The benchmark defaults to "simple" implementations, which typically indicates incomplete integration
> * **Missing Model Integration**: The transpose/quantization integration isn't fully completed
> 
> ### 4. **Text Normalization Scope Creep**
> The `TextNormalizer.swift` file contains extensive British→American spelling normalization:
> 
> ```swift
> let spellingVariations = [
>     "counselled": "counseled",
>     "travelling": "traveling",
>     // ... 50+ more mappings
> ]
> ```
> 
> **Concerns:**
> 
> * This seems excessive for ASR evaluation - most benchmarks don't require such extensive normalization
> * Could mask actual model performance issues
> * May not align with standard evaluation practices
> 
> **Question:** Do other ASR benchmarks really require this level of text normalization?
> 
> ## Performance Concerns
> ### 1. **Dynamic Duration Logic Complexity**
> The TDT decoder's duration handling is overly complex:
> 
> ```swift
> // Lines 405-421 - Complex adaptive skip logic
> internal func calculateNextTimeIndex(currentIdx: Int, skip: Int, sequenceLength: Int) -> Int {
>     let actualSkip: Int
>     if sequenceLength < 10 && skip > 2 {
>         actualSkip = 2  // Another magic number
>     } else {
>         actualSkip = min(skip, 4)  // And another one
>     }
>     return min(currentIdx + actualSkip, sequenceLength)
> }
> ```
> 
> While this achieves good benchmark results (8.1% → 7.4% WER improvement), it raises concerns about:
> 
> * Generalization to different audio types
> * Maintenance complexity
> * Debugging difficulty
> 
> ## Code Quality Issues
> ### 1. **Model File Extensions**
> ```swift
> // DownloadUtils.swift - Good change
> // Removed .mlpackage support, keeping only .mlmodelc
> ```
> 
> ✅ **Positive:** Standardizing on `.mlmodelc` is good for consistency and performance.
> 
> ### 2. **Vocabulary Integration**
> ```swift
> // AsrModels.swift:135 - Clean vocabulary loading
> logger.info("Loaded vocabulary with \(vocabulary.count) tokens")
> ```
> 
> ✅ **Positive:** Clean integration of the expanded 8193-token vocabulary.
> 
> ### 3. **Decoder State Management**
> The decoder state handling looks robust with proper initialization and fallback:
> 
> ```swift
> // AsrManager.swift:54-62 - Good error handling
> do {
>     self.microphoneDecoderState = try DecoderState()
>     self.systemDecoderState = try DecoderState()
> } catch {
>     logger.warning("Failed to create ANE-aligned decoder states, using standard allocation")
>     self.microphoneDecoderState = DecoderState(fallback: true)
>     self.systemDecoderState = DecoderState(fallback: true)
> }
> ```
> 
> ## Recommendations
> ### High Priority
> 1. **Remove magic numbers** in TDT decoder - make them configurable or derive them from audio characteristics
> 2. **Clean up beam search remnants** - remove dead code and comments
> 3. **Complete model integrations** - finish the transpose/quantization work
> 4. **Simplify text normalization** - align with standard ASR evaluation practices
> 
> ### Medium Priority
> 1. **Add comprehensive tests** for the new TDT decoder logic
> 2. **Document the dynamic duration algorithm** with clear rationale
> 3. **Add configuration options** for the hard-coded thresholds
> 
> ### Low Priority
> 1. **Profile memory usage** of the new quantized models
> 2. **Add streaming performance metrics** beyond the current benchmarks
> 
> ## Questions for Author
> 1. What's the justification for the 30-frame threshold in duration logic?
> 2. Why is such extensive text normalization needed compared to other ASR systems?
> 3. When will the encoder transpose integration be completed?
> 4. Have you tested this on audio types beyond LibriSpeech (meetings, phone calls, etc.)?
> 
> ## Overall Assessment
> The PR achieves excellent benchmark performance and successfully integrates a complex new model. However, the implementation contains concerning magic numbers and over-optimizations that may not generalize well. The code quality needs improvement before merging to production.
> 
> **Recommendation:** Request changes to address the hard-coded thresholds and complete the outstanding integrations before merge.

Remove magic numbers in TDT decoder - make them configurable or derive them from audio characteristics
- addressed
Clean up beam search remnants - remove dead code and comments
- addressed
Complete model integrations - finish the transpose/quantization work
- addressed
Simplify text normalization - align with standard ASR evaluation practices
- addressed

Medium Priority
Add comprehensive tests for the new TDT decoder logic
- addressed
Document the dynamic duration algorithm with clear rationale
Add configuration options for the hard-coded thresholds
- addressed

Low Priority
Profile memory usage of the new quantized models
Add streaming performance metrics beyond the current benchmark

BrandonWeng commented:
Closing in favor of this PR instead: https://github.com/FluidInference/FluidAudio/pull/77
